### ISSUE TITLE
Incorrect "Value '<val>' is unsubscriptable" with class method type overloading

### ISSUE THREAD

**Issue by awarrier99** [pylint-dev/pylint#9474]
### Bug description

```python
from typing import Literal, overload


class Test:
    __config: dict[str, list[int] | list[str] | list[float]]

    def __init__(self):
        self.__config = {"a": [0, 1], "b": ["abc", "def"], "c": [1.1, 2.2]}

    @overload
    def func(self, s: Literal["a"]) -> list[int]: ...
    @overload
    def func(self, s: Literal["b"]) -> list[str]: ...
    @overload
    def func(self, s: Literal["c"]) -> list[float]: ...
    def func(self, s: Literal["a"] | Literal["b"] | Literal["c"]) -> list[int] | list[str] | list[float]:
        return self.__config[s]


test = Test()
x = test.func("c")
y = x[0]
```

### Command used

```shell
pylint test.py
```

### Pylint output

```python
************* Module test
test.py:39:5: E1136: Value 'num' is unsubscriptable (unsubscriptable-object)
```

### Expected behavior

I would expect to be able to index the returned list with no linting error. It's worth noting that this same example works when the functions are defined at the module level, it seems to be only an issue within a class.

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.3
Python 3.11.5 (main, Aug 24 2023, 15:18:16) [Clang 14.0.3 (clang-1403.0.22.14.1)]
```

### OS / Environment

macOS Sonoma 14.1.1

**Comment by awarrier99** [comment]
@jacobtylerwalls ah sorry I obviously didn't search through the existing issues well enough. Thanks for the context
